### ISSUE TITLE
feat(cve): import associated CPEs when importing CVEs

### ISSUE THREAD

**Issue by d1zanv** [OpenCTI-Platform/connectors#1949]
## Use case

When you want to see the CPEs associated with a vulnerability or vice versa.

## Current Workaround

There is no workaround.

## Proposed Solution

When importing the CVEs datas into OpenCTI, included the configurations field, with the CPEs.

## Would you be willing to submit a PR?

I am willing to submit a PR if needed to help, but the last big update of the connector was made by @helene-nguyen. I think she'll be in the best position to modify her code.

**Comment by Jipegien** [comment]
Hello @d1zanv! You can submit a PR, as we are currently working on other connectors. Feel free to ask any question you may have here :)

**Comment by iFrozenPhoenix** [comment]
I'm currently working on this functionality. Currently I have it as a pure python script where all vulnerabilities are fetched from octi, then the vulnerability from the nist api is fetched and the cpe names extracted. These cpe names are then looked up in octi for their ids and the relationship is created.

Think I can convert it to an enrichment connector. @Jipegien what do you suggest, converting it to an enrichment connector, or combining the cpe and cve connector and adding this functionality?

**Comment by Jipegien** [comment]
@iFrozenPhoenix Why don't you fetch all "new" CVEs from NIST, along with the related CPEs and then create the corresponding stix data in OpenCTI. Deduplication mechanism will then take care of the matching in OpenCTI. You don't need to compare the data inside the connector logic.
It will cover the use case "I know what Softwares are impacted by CVEs"

The logic you have implemented id more adequate for an enrichment connector, but an enrichement connector works entity by entity.

**Comment by iFrozenPhoenix** [comment]
Well because there was first an cve connector that just imported the cves and scores, then an additional cpe connector was created that just imported the cpes. From this perspective it was the consequence that if one wants to relate these entities an enrichment connector is needed, or the cpe and cve connectors must be combined to 1. Therefore I asked how it should be implemented, as an enrichment connector or if the existing conectors should be combined.

**Comment by Jipegien** [comment]
I don't think it is mandatory to collect all the existing CPEs and I also think having a sort of CVE->CPE library from NIST into OpenCTI is more valuable than having enrichment of already imported CVEs. So IMHO, you should focus on improving the current CVE connector for retrieving related CPEs. Retrieval of CPE should be configurable.

Currently, I am not sure about the value of the CPE connector by itself. Why having a Software created in the platform if it is not related to a threat/vuln/IT_infra context.

**Comment by d1zanv** [comment]
> I don't think it is mandatory to collect all the existing CPEs and I also think having a sort of CVE->CPE library from NIST into OpenCTI is more valuable than having enrichment of already imported CVEs. So IMHO, you should focus on improving the current CVE connector for retrieving related CPEs. Retrieval of CPE should be configurable.
> 
> Currently, I am not sure about the value of the CPE connector by itself. Why having a Software created in the platform if it is not related to a threat/vuln/IT_infra context.

@Jipegien you are right ! For now, the CPEs doesn't have any value in OpenCTI because they can't be linked to the vulnerabilities.
When the CVE connector will be updated to import the related CPEs of a CVE, it will be a great enrichment to the platform.

**Comment by d1zanv** [comment]
I will let you update the CVE connector @iFrozenPhoenix if you already have a python script to do it. If you need any help, just let me know !

**Comment by iFrozenPhoenix** [comment]
Well one use case we use the other cpes is to map them against our infrastructure to have software asset management capabilities. Another one is that the complete dataset also contains the base cpes for the software (without version no).
While these use cases are specific to our needs someone might also have additional use cases that we currently have not in mind. Don't know if it makes sense to reduce the available knowledge that someone can use.

While thinking what the functionality that should be provided, it sounds from the characteristics that we want to enrich existing data, either by connecting a cve to a cpe or by creating and relating a cpe to a cve.

So while this functionality can also be built in the cve connector logic (if entity does not exist then create cpe and relate, else only relate) I would prefer to also keep the cpe connector, due to additional use cases that cannot be reflected with this approach.

**Comment by Jipegien** [comment]
It is a matter of priority. A lot of things are relevant enough to be done, but only one thing can be develop at a time. In my opinion, we should begin by importing the most relevant data, meaning softwares (CPEs) impacted by known vulnerabilities (CVEs), because OpenCTI is about threat management first, not assets management. For the same reason, we cannot prioritize building upon "possible" use cases. 

So I agree that it'd be great to have all these things. I am just saying, and it is just my opinion, that we should begin by the most relevant part. That's currently how we prioritize at Filigran. But we are not alone in this :) So if you have an important specific use case implying that importing all CPEs is more relevant, nothing should stop you :)

**Comment by iFrozenPhoenix** [comment]
@Jipegien I think I don't understand what you mean exactly? @d1zanv has developed the cpe connector and it is published in this repository https://github.com/OpenCTI-Platform/connectors/tree/master/external-import/cpe. I also made a minor improvement and fixed a bug. So we currently have ALL cpes and ALL cves in the platform, because they are imported by two separate connectors. To relate these two entities I wrote a script that creates the relations, because that is the missing part. This is also done, but not published as an enrichment connector here.
I asked you if I should publish my script as an Enrichtment Connector or if I should modify the CVE Connector to map the entities. So if you are talking about time, for me it is less time to provide it here for the community as an enrichment connector then looking into the code of the cve connector, implementing it and refactoring code.

**Comment by d1zanv** [comment]
> @Jipegien I think I don't understand what you mean exactly? @d1zanv has developed the cpe connector and it is published in this repository https://github.com/OpenCTI-Platform/connectors/tree/master/external-import/cpe. I also made a minor improvement and fixed a bug. So we currently have ALL cpes and ALL cves in the platform, because they are imported by two separate connectors. To relate these two entities I wrote a script that creates the relations, because that is the missing part. This is also done, but not published as an enrichment connector here. I asked you if I should publish my script as an Enrichtment Connector or if I should modify the CVE Connector to map the entities. So if you are talking about time, for me it is less time to provide it here for the community as an enrichment connector then looking into the code of the cve connector, implementing it and refactoring code.

@Jipegien For me, it should be implement in the CVE connector as the data you want to use for your enrichment connector are coming from the exact same source as the CVE connector. It's just that when the CVE connector was developped, there was no way of adding the CPEs to OpenCTI, so the dev had no reason to extract the CPEs from the CVE API.

**Comment by Jipegien** [comment]
Sorry for the misunderstanding. Do what you think is the most relevant for you. On my side, I don't currently see the value of importing all CPEs in the platform and use a connector to do so. The CPE connector is not deployed in our instances currently, thus I see more value to improve the CVE connector because it is the use case I see as important.

**Comment by d1zanv** [comment]
> Sorry for the misunderstanding. Do what you think is the most relevant for you. On my side, I don't currently see the value of importing all CPEs in the platform and use a connector to do so. The CPE connector is not deployed in our instances currently, thus I see more value to improve the CVE connector because it is the use case I see as important.

@Jipegien If I understand you correctly, you would have just updated the CVE connector, so that when a CPE is find related to a CVE, you create it and add a relation between the 2 ?

**Comment by Jipegien** [comment]
Yes

**Comment by iFrozenPhoenix** [comment]
@d1zanv I created a repo under my account with the current status. You can use, copy, modify,... it as needed. https://github.com/iFrozenPhoenix/octi-cve-enrichment
Thank you for creating the cpe connector 👍️

**Comment by iFrozenPhoenix** [comment]
@Jipegien I wanted to come back to this issue, also calmed down, sorry for the language.
Do you still think it is valuable to eliminate the CPE connector and integrate into the CVE connector? I ask it again because your company currently focuses on vulnerability management and asset management integration (tenable I think).
Further I think it would be valuable to see the lifecycle of a vulnerability (i.e. disclosed, proof of concept, exploited). I wanted to discuss how all these things could be integrated best into the platform. Because I really don't know what would be best. Single Connector with options or multiple connectors. Would be cool to discuss these things and find the best solution that fits all our needs. Regards

**Comment by iFrozenPhoenix** [comment]
Additions: Currently I add the lifecycle state also via a script with data from explitdb.

**Comment by Jipegien** [comment]
I still think there is more value to import CPE and Software IF they are Threat related or "Infrastructure" related. In respect with that, i see more value in the CVE connector (importing Software that are actually linked to CVE) and connector like Tanium Asset management (importing Software part of the Infrastructure the user must protect). CPE connector will import every known Softwares (my understanding is it will include Softwares not related to CVE, but perhaps I'm wrong here). I think it will result to a lot of unrelevant objects in the platform (because not related to threats or infrastructures to protect). So, It is not about eliminate CPE connector, it is more a priority matter.

But again, if it is more relevant to integrate CPE connector for your specific use case, I see no harm in developing it. It is just not a priority on our side :)

I also see a lot of value on everything that could help assess how a vulnerability is exploited in the wild (EPSS and so on)

**Comment by iFrozenPhoenix** [comment]
Ok, then I will open a PR at the weekend for the CPE connector that adds a exploit label at exploited label on exploited vulnerabilities (my current approach). I would make this optional and configurable. Is this in your sense?


Mit freundlichen Grüßen / Kind regards
________________________________

**Comment by d1zanv** [comment]
> Ok, then I will open a PR at the weekend for the CPE connector that adds a exploit label at exploited label on exploited vulnerabilities (my current approach). I would make this optional and configurable. Is this in your sense? Mit freundlichen Grüßen / Kind regards

Do you need my help @iFrozenPhoenix ?

**Comment by V1D1AN** [comment]
Hi the teams, I have try your enrichment script python for make the association between CVE and CPE.
I have do the request https://github.com/OpenCTI-Platform/opencti/issues/7200 for the same thing but i didn't see you have already the same idea.

I have try to make a connector-enrichment with the template of OpenCTI.
I have always error 404 when i add the NIST_API_KEY

**Comment by damians-filigran** [comment]
Just adding to this, as I came across the same requirement.

I see that NVD publish the CVE stream, the CPE stream, both separately, and _then_ the CPE_match stream, which maps CPEs to CVEs.
https://nvd.nist.gov/vuln/data-feeds#cpeMatch

An example REST GET of the stream is here
https://services.nvd.nist.gov/rest/json/cpematch/2.0

My understanding is that what we are missing is a third connector for this match stream, to link the two?

Since the match stream seems to lag the CPE/CVE streams, it seems a simplest implementation is to simply run this separately assuming that the CPE and CVE feeds have already created the entities.


Furthermore, in order to avoid importing the entire 1.3m current CPE entries, I would propose that the CPE connector be modified to support the [keywordsearch](https://nvd.nist.gov/developers/products) parameter, which can take simple search strings to find matching CPEs. If this is run on a polling interval, with a timerange since the GET, and takes an input parameter of a CSV list of software product keywords which it iterates through, looking for any new matches each poll, then this would allow the CPE connector to only pull CPEs relating to software that is in a 'watchlist' for the user.

**Comment by romain-filigran** [comment]
Implemented in #6012, released in **7.260430.0**.

The connector now resolves affected CPEs via the NVD CPE Match API and imports them as **Software** SCOs with a `has` relationship to the Vulnerability — the canonical relationship type for Software ↔ Vulnerability in OpenCTI.

Opt-in via `CVE_IMPORT_SOFTWARE` (default `false`), with `CVE_CPE_HISTORY_INTERVAL` and `CVE_CPE_MAX_CONCURRENCY` to bound the ingested volume. See the [config documentation](https://github.com/OpenCTI-Platform/connectors/blob/master/external-import/cve/__metadata__/CONNECTOR_CONFIG_DOC.md).

Closing as completed.
